### PR TITLE
#1401. [Patterns] If case statement and element tests added

### PR DESCRIPTION
**Pull request by sgrekhov**

### CODE REVIEW

**Comment by eernstg** [comment]
@chloestefantsova, would it be possible for you to take a look at this one? I'm out next week... ;-)

File: LanguageFeatures/Patterns/if_case_A02_t02.dart [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+        testParenthesized1(40, 42));
+    Expect.mapEquals({"key1": 1, "key2": 2, "key4": 4},
+        testParenthesized2(42));
+    Expect.mapEquals({"key1": 1, "key3": 3, "key4": 4},
+        testParenthesized2(40));
+    Expect.mapEquals({"key1": 1, "key2": 2, "key4": 4}, testCast(42));
+    Expect.mapEquals({"key1": 1, "key3": 3, "key4": 4}, testCast(41));
+    Expect.mapEquals({"key1": 1, "key2": 2, "key4": 4}, testNullCheck1(1));
+    Expect.mapEquals({"key1": 1, "key3": 3, "key4": 4}, testNullCheck1(-1));
+    Expect.mapEquals({"key1": 1, "key3": 3, "key4": 4}, testNullCheck1(null));
```
Comment (eernstg) [inline LanguageFeatures/Patterns/if_case_A02_t02.dart:244]: Missing `testNullCheck2` cases?

File: LanguageFeatures/Patterns/if_case_A02_t02.dart [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+        testMap2({"key1": 2, "key2": -2}));
+    Expect.mapEquals({"key1": 1, "key3": 3, "key4": 4},
+        testMap2({"key1": 1, "key2": 2}));
+    Expect.mapEquals({"key1": 1, "key2": 2, "key4": 4},
+        testMap2({"key1": 1, "key2": -2, "key3": 3}));
+    Expect.mapEquals({"key1": 1, "key2": 2, "key4": 4}, testRecord1((1, 2)));
+    Expect.mapEquals({"key1": 1, "key3": 3, "key4": 4}, testRecord1((2, 2)));
+    Expect.mapEquals({"key1": 1, "key3": 3, "key4": 4}, testRecord1((1, -2)));
+    Expect.mapEquals({"key1": 1, "key2": 2, "key4": 4},
+        testRecord2((1, 2, n: 1)));
```
Comment (eernstg) [inline LanguageFeatures/Patterns/if_case_A02_t02.dart:282]:
```suggestion
        testRecord2((1, 2, n: -1)));
```

File: LanguageFeatures/Patterns/if_case_A02_t02.dart [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+    Expect.mapEquals({"key1": 1, "key3": 3, "key4": 4},
+        testRecord2((2, 2, n: 1)));
+    Expect.mapEquals({"key1": 1, "key3": 3, "key4": 4},
+        testRecord2((1, 2, n: -1)));
```
Comment (eernstg) [inline LanguageFeatures/Patterns/if_case_A02_t02.dart:286]:
```suggestion
        testRecord2((1, 2, n: 1)));
```

File: LanguageFeatures/Patterns/if_case_A02_t02.dart [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+    Expect.mapEquals({"key1": 1, "key3": 3, "key4": 4},
+        testRecord2((1, n: 1)));
+    Expect.mapEquals({"key1": 1, "key3": 3, "key4": 4},
+        testRecord2((1, 2, n: 1, n2: 2)));
+    Expect.mapEquals({"key1": 1, "key2": 2, "key4": 4}, testObject1(Square(3)));
+    Expect.mapEquals({"key1": 1, "key3": 3, "key4": 4}, testObject1(Circle(3)));
+    Expect.mapEquals({"key1": 1, "key3": 3, "key4": 4}, testObject1(Square(1)));
+    Expect.mapEquals({"key1": 1, "key2": 2, "key4": 4}, testObject1(Circle(3)));
```
Comment (eernstg) [inline LanguageFeatures/Patterns/if_case_A02_t02.dart:294]:
```suggestion
    Expect.mapEquals({"key1": 1, "key2": 2, "key4": 4}, testObject2(Circle(3)));
```

Same fixes again in lines below.

File: LanguageFeatures/Patterns/if_case_A02_t03.dart [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+    Expect.setEquals({1, 2, 4}, testParenthesized1(40, 40));
+    Expect.setEquals({1, 3, 4}, testParenthesized1(42, 40));
+    Expect.setEquals({1, 3, 4}, testParenthesized1(40, 42));
+    Expect.setEquals({1, 2, 4}, testParenthesized2(42));
+    Expect.setEquals({1, 3, 4}, testParenthesized2(40));
+    Expect.setEquals({1, 2, 4}, testCast(42));
+    Expect.setEquals({1, 3, 4}, testCast(41));
+    Expect.setEquals({1, 2, 4}, testNullCheck1(1));
+    Expect.setEquals({1, 3, 4}, testNullCheck1(-1));
+    Expect.setEquals({1, 3, 4}, testNullCheck1(null));
```
Comment (eernstg) [inline LanguageFeatures/Patterns/if_case_A02_t03.dart:228]: Missing `testNullCheck2` cases?

File: LanguageFeatures/Patterns/if_case_A02_t03.dart [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+    Expect.setEquals({1, 3, 4}, testMap1({"key1": 1, "key2": -2}));
+    Expect.setEquals({1, 3, 4}, testMap1({"key1": 1, "key2": 2, "key3": 3}));
+    Expect.setEquals({1, 2, 4}, testMap2({"key1": 1, "key2": -2}));
+    Expect.setEquals({1, 3, 4}, testMap2({"key1": 2, "key2": -2}));
+    Expect.setEquals({1, 3, 4}, testMap2({"key1": 1, "key2": 2}));
+    Expect.setEquals({1, 2, 4}, testMap2({"key1": 1, "key2": -2, "key3": 3}));
+    Expect.setEquals({1, 2, 4}, testRecord1((1, 2)));
+    Expect.setEquals({1, 3, 4}, testRecord1((2, 2)));
+    Expect.setEquals({1, 3, 4}, testRecord1((1, -2)));
+    Expect.setEquals({1, 2, 4}, testRecord2((1, 2, n: 1)));
```
Comment (eernstg) [inline LanguageFeatures/Patterns/if_case_A02_t03.dart:257]:
```suggestion
    Expect.setEquals({1, 2, 4}, testRecord2((1, 2, n: -1)));
```

File: LanguageFeatures/Patterns/if_case_A02_t03.dart [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+    Expect.setEquals({1, 3, 4}, testRecord2((2, 2, n: 1)));
+    Expect.setEquals({1, 3, 4}, testRecord2((1, 2, n: -1)));
```
Comment (eernstg) [inline LanguageFeatures/Patterns/if_case_A02_t03.dart:259]:
```suggestion
    Expect.setEquals({1, 3, 4}, testRecord2((1, 2, n: 1)));
```

File: LanguageFeatures/Patterns/if_case_A02_t03.dart [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+    Expect.setEquals({1, 3, 4}, testRecord2((1, n: 1)));
+    Expect.setEquals({1, 3, 4}, testRecord2((1, 2, n: 1, n2: 2)));
+    Expect.setEquals({1, 2, 4}, testObject1(Square(3)));
+    Expect.setEquals({1, 3, 4}, testObject1(Circle(3)));
+    Expect.setEquals({1, 3, 4}, testObject1(Square(1)));
+    Expect.setEquals({1, 2, 4}, testObject1(Circle(3)));
```
Comment (eernstg) [inline LanguageFeatures/Patterns/if_case_A02_t03.dart:265]:
```suggestion
    Expect.setEquals({1, 2, 4}, testObject2(Circle(3)));
```
Same corrections again in the following lines.

**Review by eernstg — COMMENTED** [review]
A handful of glitches encountered.

**Comment by sgrekhov** [comment]
Very detailed, thank you! Now all of the tests in this PR works in `dartk`

**Review by eernstg — APPROVED** [review]
LGTM

**Pull request merged**